### PR TITLE
[pt2e][qat] Support conv-transpose-bn[-relu] QAT fusion

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123652

Summary: This commit adds support for QAT fusion for the
[conv-transpose-bn] and [conv-transpose-bn-relu] patterns.

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn1d.test_qat_conv_transpose_bn
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn1d.test_qat_conv_transpose_bn_relu
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d.test_qat_conv_transpose_bn
python test/test_quantization.py TestQuantizePT2EQAT_ConvBn2d.test_qat_conv_transpose_bn_relu

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: https://github.com/pytorch/pytorch/issues/122224

Differential Revision: [D55930704](https://our.internmc.facebook.com/intern/diff/D55930704)

### CODE REVIEW

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 1]
```diff
@@ -347,6 +353,63 @@ def _verify_symmetric_xnnpack_qat_graph_helper(
...
+        )
+        self.assertEqual(
+            output_q_node.target,
+            torch.ops.quantized_decomposed.quantize_per_tensor.default,
+        )
+        self.assertTrue(_is_conv_node(conv_node))
+        self.assertEqual(list(output_dq_node.args[-3:]), [-128, 127, torch.int8])
+        self.assertEqual(list(output_q_node.args[-3:]), [-128, 127, torch.int8])
+
+        # Verify: conv_weight - weight_dq - conv
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:384]: for these can we use existing self.checkGraphModuleNodes or something like that to check?

File: torch/ao/quantization/pt2e/qat_utils.py [hunk 2]
```diff
@@ -522,11 +522,12 @@ def _fuse_conv_bn_qat(m: GraphModule) -> GraphModule:
     has_bn = any(_is_bn_node(n) for n in m.graph.nodes)
     if not has_bn:
         return m
-    m = _fuse_conv_bn_qat_helper(m, F.conv1d, _conv1d_bn_example_inputs, is_cuda=False)
-    m = _fuse_conv_bn_qat_helper(m, F.conv2d, _conv2d_bn_example_inputs, is_cuda=False)
-    if torch.cuda.is_available():
-        m = _fuse_conv_bn_qat_helper(m, F.conv1d, _conv1d_bn_example_inputs, is_cuda=True)
-        m = _fuse_conv_bn_qat_helper(m, F.conv2d, _conv2d_bn_example_inputs, is_cuda=True)
+    is_cuda_options = [True, False] if torch.cuda.is_available() else [False]
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/qat_utils.py:525]: one question here is do we want to separate model preparation, v.s. model execution? right now it seems that how the model is prepared depends on the status of whether cuda is available or not when we run the model preparation code, which doesn't make a lot of sense to me. I'm wondering whether we need to make this decision more explicit for users here.

File: torch/ao/quantization/pt2e/utils.py [hunk 3]
```diff
@@ -174,14 +174,15 @@ def _is_conv_node(n: Node):
     return n.op == "call_function" and n.target in [
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/utils.py:174]: nit: probably need some renaming here, since `_is_conv_node` is actually `_is_conv_or_conv_transpose_node` now

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py (hunk 1 above)
Comment (andrewor14) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:384]: Ok let me try. I think we can do this for the convert graph since it's simpler. For the prepare graph we probably still need the existing code

File: torch/ao/quantization/pt2e/qat_utils.py (hunk 2 above)
Comment (andrewor14) [reply torch/ao/quantization/pt2e/qat_utils.py:525]: This is currently needed because we have different batch norm ops depending on cuda vs cpu, so we need to match both versions. After batch norm consolidation we might not need this anymore. However, I'm hesitant to make the general assumption that batch norm is the only op we need to worry about. In general, export gives us a graph specialized to the device, and I feel there's not much we can do about that.

Comment (jerryzh168) [reply torch/ao/quantization/pt2e/qat_utils.py:525]: so my point is model preparation/quantization should be decoupled from the whether the model happens to be prepared/quantized in cuda or cpu right?

Comment (andrewor14) [reply torch/ao/quantization/pt2e/qat_utils.py:525]: I agree at a high level. The way we get the cuda version here is by calling `example_inputs.cuda()` first before exporting. I don't think that will work on a CPU-only machine. Also this is just refactoring here so if we decide to change this let's do it in a separate PR

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

File: test/quantization/pt2e/test_quantize_pt2e_qat.py [hunk 4]
```diff
@@ -347,12 +358,72 @@ def _verify_symmetric_xnnpack_qat_graph_helper(
...
 
 class TestQuantizePT2EQAT_ConvBn_Base(PT2EQATTestCase):
     """
     Base TestCase to be used for all conv-bn[-relu] fusion patterns.
     """
 
+    # TODO: how can we avoid adding every new test to dynamo/expected_test_failures?
+    # Otherwise it fails with the following error:
+    #   torch._dynamo.exc.InternalTorchDynamoError:
+    #   'QuantizationConfig' object has no attribute '__bool__'
```
Comment (jerryzh168) [inline test/quantization/pt2e/test_quantize_pt2e_qat.py:425]: can we just add `__bool__` to QuantizationConfig?

File: torch/ao/quantization/pt2e/qat_utils.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/qat_utils.py:528]: oh I see, sounds good

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py [hunk 5]
```diff
@@ -344,21 +346,11 @@ def _do_annotate_conv_relu(
...
-                torch.ops.aten.conv1d.default,
-                torch.ops.aten.conv2d.default,
-            ]
+
         if (
             not isinstance(maybe_conv_node, Node)
-            or maybe_conv_node.op != "call_function"
-            or maybe_conv_node.target not in conv_ops
+            or (is_conv_transpose and not _is_conv_transpose_node(maybe_conv_node))
+            or (not is_conv_transpose and not _is_conv_node(maybe_conv_node))
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:353]: this looks a bit complicated, can this be simplified? maybe remove the `is_conv_transpose` flag and just pass in a function to check the node?

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks for addressing the comments, left some nits

File: test/quantization/pt2e/test_quantize_pt2e_qat.py (hunk 4 above)
Comment (andrewor14) [reply test/quantization/pt2e/test_quantize_pt2e_qat.py:425]: Not sure. It's an existing failure so I'll debug separately

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55930704).
